src/cockpit/config.py

```diff
@@ -15,10 +15,61 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <https://www.gnu.org/licenses/>.
 
+import configparser
+import logging
 import os
 
 from pathlib import Path
 
+from systemd_ctypes import bus
+
+logger = logging.getLogger(__name__)
+
 ETC_COCKPIT = Path('/etc/cockpit')
 XDG_CONFIG_HOME = Path(os.getenv('XDG_CONFIG_HOME') or os.path.expanduser('~/.config'))
 DOT_CONFIG_COCKPIT = XDG_CONFIG_HOME / 'cockpit'
+
+
+class Config(bus.Object, interface='cockpit.Config'):
+    def __init__(self):
+        self.reload()
+
+    @bus.Interface.Method(out_types='s', in_types='ss')
+    def get_string(self, section, key):
+        try:
+            return self.config[section][key]
+        except KeyError:
+            raise bus.BusError('cockpit.Config.KeyError', f'key {key} in section {section} does not exist')
+
+    @bus.Interface.Method(out_types='u', in_types='ssuuu')
+    def get_u_int(self, section, key, default, maximum, minimum):
+        try:
+            value = self.config[section][key]
+        except KeyError:
+            return default
+
+        try:
+            int_val = int(value)
+        except ValueError:
+            logger.warning(f'cockpit.conf: [{section}] {key} is not an integer')
+            return default
+
+        return min(max(int_val, minimum), maximum)
+
+    @bus.Interface.Method()
+    def reload(self):
+        self.config = configparser.ConfigParser(interpolation=None)
+
+        config_dirs = os.environ.get('XDG_CONFIG_DIRS', '/etc').split(':')
+        for config_dir in config_dirs:
+            config_file = os.path.join(config_dir, 'cockpit', 'cockpit.conf')
+            if os.path.exists(config_file):
+                logger.debug("cockpit.Config: loading %s", config_file)
+                self.config.read(config_file)
+                break
+
+        # it's ok to not have a config file and thus leave self.config empty
+
+
+class Environment(bus.Object, interface='cockpit.Environment'):
+    variables = bus.Interface.Property('a{ss}', value=os.environ)
```

Comment:
Good point, done. If we accept the environment as malleable, then it would actually be cleaner to have a property getter function that returns the current os.environ at the time of calling Get() -- but AFAICS systemd_ctype's API does not have a concept of a property getter function, or I'm not seeing it?

Anyway, we only use this endpoint in two unit tests, and changing the environment of a running process is fragile at best in Linux, so I suppose right now we don't need to go overboard with that.